File: src/Sample/Plumix.Sample/Demos/Material/VisualDensityDemoPage.cs
using System;
using System.Collections.Generic;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/material/visual_density_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class VisualDensityDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new VisualDensityDemoPageState();
    }
}

internal sealed class VisualDensityDemoPageState : State
{
    private static readonly IReadOnlyList<(string Name, VisualDensity Density)> Profiles =
    [
        ("standard", VisualDensity.Standard),
        ("comfortable", VisualDensity.Comfortable),
        ("compact", VisualDensity.Compact),
        ("(3, 3)", new VisualDensity(horizontal: 3, vertical: 3)),
        ("(-3, -3)", new VisualDensity(horizontal: -3, vertical: -3)),
    ];

    private static readonly IReadOnlyList<TargetPlatform> Platforms =
    [
        TargetPlatform.Android,
        TargetPlatform.IOS,
        TargetPlatform.Fuchsia,
        TargetPlatform.Linux,
        TargetPlatform.MacOS,
        TargetPlatform.Windows,
    ];

    private int _profileIndex;
    private int _platformIndex;

    public override Widget Build(BuildContext context)
    {
        (string name, VisualDensity density) = Profiles[_profileIndex];
        TargetPlatform platform = Platforms[_platformIndex];
        var platformTheme = new ThemeData(platform: platform);

        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("VisualDensity + platform defaults", fontSize: 20, color: Colors.Black),
                new Text(
                    "Density is unitless: one unit is four logical pixels per axis. ThemeData takes "
                    + "its default from the theme's platform, not the host.",
                    fontSize: 14,
                    color: Color.Parse("#8A000000")),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildControlButton(
                            label: $"density: {name}",
                            onTap: () => SetState(() =>
                                _profileIndex = (_profileIndex + 1) % Profiles.Count),
                            width: 180,
                            background: Color.Parse("#FFE9F0FF")),
                        BuildControlButton(
                            label: $"platform: {platform}",
                            onTap: () => SetState(() =>
                                _platformIndex = (_platformIndex + 1) % Platforms.Count),
                            width: 180,
                            background: Color.Parse("#FFEAF6F7")),
                    ]),
                new Expanded(
                    child: new SingleChildScrollView(
                        child: new Column(
                            crossAxisAlignment: CrossAxisAlignment.Stretch,
                            spacing: 14,
                            children:
                            [
                                BuildPlatformDefaults(platform, platformTheme),
                                BuildDensityFacts(density),
                                BuildSizedProbes(density),
                            ]))),
            ]);
    }

    private Widget BuildPlatformDefaults(TargetPlatform platform, ThemeData theme)
    {
        return BuildCard(
            "ThemeData(platform: " + platform + ") defaults",
            [
                BuildFactRow("visualDensity", theme.VisualDensity.ToString()),
                BuildFactRow("materialTapTargetSize", theme.MaterialTapTargetSize.ToString()),
                BuildFactRow(
                    "VisualDensity.DefaultDensityForPlatform",
                    VisualDensity.DefaultDensityForPlatform(platform).ToString()),
            ]);
    }

    private Widget BuildDensityFacts(VisualDensity density)
    {
        Vector adjustment = density.BaseSizeAdjustment;
        BoxConstraints effective = density.EffectiveConstraints(
            BoxConstraints.TightFor(width: 48, height: 48));
        return BuildCard(
            $"{density}",
            [
                BuildFactRow("baseSizeAdjustment", $"({adjustment.X:0.#}, {adjustment.Y:0.#})"),
                BuildFactRow(
                    "effectiveConstraints(48x48)",
                    $"min {effective.MinWidth:0.#}x{effective.MinHeight:0.#}, "
                    + $"max {effective.MaxWidth:0.#}x{effective.MaxHeight:0.#}"),
            ]);
    }

    private Widget BuildSizedProbes(VisualDensity density)
    {
        List<Widget> probes = [];
        foreach ((string name, VisualDensity profile) in Profiles)
        {
            bool selected = profile == density;
            probes.Add(new Padding(
                insets: new Thickness(0, 0, 0, 8),
                child: new Row(
                    spacing: 10,
                    children:
                    [
                        new SizedBox(
                            width: 110,
                            child: new Text(
                                name,
                                fontSize: 12,
                                color: selected ? Colors.Black : Color.Parse("#FF607D8B"))),
                        new Theme(
                            data: ThemeData.Light with { VisualDensity = profile },
                            child: new ElevatedButton(
                                onPressed: () => { },
                                child: new Text("Button", fontSize: 12))),
                    ])));
        }

        return BuildCard("The same button under each density", probes);
    }

    private Widget BuildCard(string title, IReadOnlyList<Widget> children)
    {
        List<Widget> rows = [new Text(title, fontSize: 14, color: Colors.Black)];
        foreach (Widget child in children)
        {
            rows.Add(child);
        }

        return new Container(
            color: Color.Parse("#FFF7F9FC"),
            padding: new Thickness(12),
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 8,
                children: rows));
    }

    private Widget BuildFactRow(string label, string value)
    {
        return new Row(
            spacing: 10,
            children:
            [
                new Expanded(child: new Text(label, fontSize: 12, color: Colors.Black)),
                new Text(value, fontSize: 12, color: Color.Parse("#FF607D8B")),
            ]);
    }

    private Widget BuildControlButton(
        string label,
        Action onTap,
        double width,
        Color background)
    {
        return new SizedBox(
            width: width,
            child: new TextButton(
                onPressed: onTap,
                child: new Text(label, fontSize: 12),
                style: TextButton.StyleFrom(
                    foregroundColor: Colors.Black,
                    backgroundColor: background,
                    padding: new Thickness(10, 8),
                    minimumSize: new Size(64, 36),
                    shape: new RoundedRectangleBorder(
                        borderRadius: BorderRadius.Circular(8)))));
    }
}

File: dart_sample/lib/demos/material/visual_density_demo_page.dart
import 'package:material_ui/material_ui.dart';

class VisualDensityDemoPage extends StatefulWidget {
  const VisualDensityDemoPage({super.key});

  @override
  State<VisualDensityDemoPage> createState() => _VisualDensityDemoPageState();
}

class _VisualDensityDemoPageState extends State<VisualDensityDemoPage> {
  static const List<(String, VisualDensity)> _profiles = <(String, VisualDensity)>[
    ('standard', VisualDensity.standard),
    ('comfortable', VisualDensity.comfortable),
    ('compact', VisualDensity.compact),
    ('(3, 3)', VisualDensity(horizontal: 3, vertical: 3)),
    ('(-3, -3)', VisualDensity(horizontal: -3, vertical: -3)),
  ];

  static const List<TargetPlatform> _platforms = <TargetPlatform>[
    TargetPlatform.android,
    TargetPlatform.iOS,
    TargetPlatform.fuchsia,
    TargetPlatform.linux,
    TargetPlatform.macOS,
    TargetPlatform.windows,
  ];

  int _profileIndex = 0;
  int _platformIndex = 0;

  @override
  Widget build(BuildContext context) {
    final (String name, VisualDensity density) = _profiles[_profileIndex];
    final TargetPlatform platform = _platforms[_platformIndex];
    final ThemeData platformTheme = ThemeData(platform: platform);

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'VisualDensity + platform defaults',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Density is unitless: one unit is four logical pixels per axis. ThemeData takes '
          "its default from the theme's platform, not the host.",
          style: TextStyle(fontSize: 14, color: Color(0x8A000000)),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildControlButton(
              label: 'density: $name',
              onTap: () => setState(
                () => _profileIndex = (_profileIndex + 1) % _profiles.length,
              ),
              width: 180,
              background: const Color(0xFFE9F0FF),
            ),
            _buildControlButton(
              label: 'platform: ${platform.name}',
              onTap: () => setState(
                () => _platformIndex = (_platformIndex + 1) % _platforms.length,
              ),
              width: 180,
              background: const Color(0xFFEAF6F7),
            ),
          ],
        ),
        Expanded(
          child: SingleChildScrollView(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              spacing: 14,
              children: <Widget>[
                _buildPlatformDefaults(platform, platformTheme),
                _buildDensityFacts(density),
                _buildSizedProbes(density),
              ],
            ),
          ),
        ),
      ],
    );
  }

  Widget _buildPlatformDefaults(TargetPlatform platform, ThemeData theme) {
    return _buildCard('ThemeData(platform: ${platform.name}) defaults', <Widget>[
      _buildFactRow('visualDensity', theme.visualDensity.toString()),
      _buildFactRow('materialTapTargetSize', theme.materialTapTargetSize.name),
      _buildFactRow(
        'VisualDensity.defaultDensityForPlatform',
        VisualDensity.defaultDensityForPlatform(platform).toString(),
      ),
    ]);
  }

  Widget _buildDensityFacts(VisualDensity density) {
    final Offset adjustment = density.baseSizeAdjustment;
    final BoxConstraints effective = density.effectiveConstraints(
      BoxConstraints.tightFor(width: 48, height: 48),
    );
    return _buildCard('$density', <Widget>[
      _buildFactRow(
        'baseSizeAdjustment',
        '(${adjustment.dx.toStringAsFixed(0)}, ${adjustment.dy.toStringAsFixed(0)})',
      ),
      _buildFactRow(
        'effectiveConstraints(48x48)',
        'min ${effective.minWidth.toStringAsFixed(0)}x${effective.minHeight.toStringAsFixed(0)}, '
            'max ${effective.maxWidth.toStringAsFixed(0)}x${effective.maxHeight.toStringAsFixed(0)}',
      ),
    ]);
  }

  Widget _buildSizedProbes(VisualDensity density) {
    final List<Widget> probes = <Widget>[];
    for (final (String name, VisualDensity profile) in _profiles) {
      final bool selected = profile == density;
      probes.add(
        Padding(
          padding: const EdgeInsets.only(bottom: 8),
          child: Row(
            spacing: 10,
            children: <Widget>[
              SizedBox(
                width: 110,
                child: Text(
                  name,
                  style: TextStyle(
                    fontSize: 12,
                    color: selected ? Colors.black : const Color(0xFF607D8B),
                  ),
                ),
              ),
              Theme(
                data: ThemeData.light().copyWith(visualDensity: profile),
                child: ElevatedButton(
                  onPressed: () {},
                  child: const Text('Button', style: TextStyle(fontSize: 12)),
                ),
              ),
            ],
          ),
        ),
      );
    }

    return _buildCard('The same button under each density', probes);
  }

  Widget _buildCard(String title, List<Widget> children) {
    return Container(
      color: const Color(0xFFF7F9FC),
      padding: const EdgeInsets.all(12),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        spacing: 8,
        children: <Widget>[
          Text(title, style: const TextStyle(fontSize: 14, color: Colors.black)),
          ...children,
        ],
      ),
    );
  }

  Widget _buildFactRow(String label, String value) {
    return Row(
      spacing: 10,
      children: <Widget>[
        Expanded(
          child: Text(label, style: const TextStyle(fontSize: 12, color: Colors.black)),
        ),
        Text(value, style: const TextStyle(fontSize: 12, color: Color(0xFF607D8B))),
      ],
    );
  }

  Widget _buildControlButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: TextButton(
        onPressed: onTap,
        style: TextButton.styleFrom(
          foregroundColor: Colors.black,
          backgroundColor: background,
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
          minimumSize: const Size(64, 36),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
        ),
        child: Text(label, style: const TextStyle(fontSize: 12)),
      ),
    );
  }
}
